<a href="https://colab.research.google.com/github/LeeMonet/TTIC31230/blob/main/PS2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Instructions

For this assignment you will use PyTorch instead of EDF to implement and train neural networks. The experiments in this assignment will take a long time to run without a GPU, but you can run the notebook remotely on Google Colab and have access to GPUs for free -- in this case you don't have to worry about installing PyTorch as it is available by default in Google Colab's environment.

In case you will be running the experiments in your own machine, you should install PyTorch -- there are multiple tutorials online and it is especially easy if you're using Anaconda. Check https://pytorch.org/tutorials/ for some PyTorch tutorials -- this assignment assumes that you know the basics like defining models with multiple modules and coding up functions to train models with PyTorch optimizers. To 

To use Google Colab, you should access https://colab.research.google.com/ and upload this notebook to your workspace. To use a GPU, go to Edit -> Notebook settings and select GPU as the accelerator.

Unlike previous assignments, in this one you will have to do some writing instead of just coding. Try to keep your answers short and precise, and you are encouraged to write equations if needed (you can do that using markdown cells). You can also use code as part of your answers (like plotting and printing, etc). Blue text indicates questions or things that you should discuss/comment, and there will red "ANSWER (BEGIN)" and "ANSWER (END)" markdown cells to indicate that you should add cells with your writeup between these two. **Make sure not to redefine variables or functions in your writeup, which can change the behavior of the next cells.**

Finally, you might have to do minor changes to the provided code due to differences in python/pytorch versions. You can post on piazza if there's a major, non-trivial change that you had to do (so other students can be aware of it and how to proceed), but for minor changes you should just apply them and keep working on the assignment.

In [1]:
import torch, math, copy
import numpy as np
from torchvision import datasets, transforms
import torch.nn as nn
import torch.nn.init as init
import torch.nn.functional as F

In [2]:
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))])

train_dataset = datasets.MNIST("data", train=True, download=True, transform=transform)
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=256, shuffle=True)

test_dataset = datasets.MNIST("data", train=False, download=True, transform=transform)
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=256, shuffle=False)

  0%|          | 0/9912422 [00:00<?, ?it/s]

Extracting data/MNIST/raw/train-images-idx3-ubyte.gz to data/MNIST/raw



  0%|          | 0/28881 [00:00<?, ?it/s]

Extracting data/MNIST/raw/train-labels-idx1-ubyte.gz to data/MNIST/raw



  0%|          | 0/1648877 [00:00<?, ?it/s]

Extracting data/MNIST/raw/t10k-images-idx3-ubyte.gz to data/MNIST/raw



  0%|          | 0/4542 [00:00<?, ?it/s]

Extracting data/MNIST/raw/t10k-labels-idx1-ubyte.gz to data/MNIST/raw



/usr/local/lib/python3.7/dist-packages/torchvision/datasets/mnist.py:498: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


Fill the missing code below. In both train_epoch and test, total_correct should be the total number of correctly classified samples, while total_samples should be the total number of samples that have been iterated over.

In [3]:
def train(epochs, model, criterion, optimizer, train_loader, test_loader):
    for epoch in range(epochs):
        train_err = train_epoch(model, criterion, optimizer, train_loader)
        test_err = test(model, test_loader)
        print('Epoch {:03d}/{:03d}, Train Error {:.2f}% || Test Error {:.2f}%'.format(epoch, epochs, train_err*100, test_err*100))
    return train_err, test_err
    
def train_epoch(model, criterion, optimizer, loader):
    total_correct = 0.
    total_samples = 0.
    
    for batch_idx, (data, target) in enumerate(loader):
        if torch.cuda.is_available():
            data, target = data.cuda(), target.cuda()

        # insert code to feed the data to the model and collect its output
        output = model(data)

        # insert code to compute the loss from output and the true target
        loss = criterion(output, target)

        # insert code to update total_correct and total_samples
        # total_correct: total number of correctly classified samples
        # total_samples: total number of samples seen so far
        total_correct += torch.eq(torch.max(output, 1).indices, target).long().sum().item()



        #output = output.detach().numpy()
        #target = target.numpy()
        #pred = np.argmax(output, axis=1)
        #total_correct += (pred==target).astype(int).sum().item()
        total_samples += data.size(0)


        # insert code to update the parameters using optimizer
        # be careful in this part as an incorrect implementation will affect
        # all your experiments and have a significant impact on your grade!
        # in particular, note that pytorch does --not-- automatically
        # clear the parameter's gradients: check tutorials to see
        # how this can be done with a single method call.
        ...
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    return 1 - total_correct/total_samples
    
def test(model, loader):
    total_correct = 0.
    total_samples = 0.
    model.eval()
    
    with torch.no_grad():
        for batch_idx, (data, target) in enumerate(loader):
            if torch.cuda.is_available():
                data, target = data.cuda(), target.cuda()

            # insert code to feed the data to the model and collect its output
            output = model(data)

            # insert code to update total_correct and total_samples
            # total_correct: total number of correctly classified samples
            # total_samples: total number of samples seen so far
            
            total_correct += torch.eq(torch.max(output, 1).indices, target).long().sum().item()
            #output = output.detach().numpy() # make numpy array, remove grad
            #target = target.numpy()
            #pred = np.argmax(output, axis=1) #find index of highest value
            #total_correct += (pred==target).astype(int).sum().item()

            total_samples += data.size(0)


    return 1 - total_correct/total_samples

### CNN with Tanh activations

Next, you should implement a baseline model so you can check how increasing the number of layers can make a network considerably harder to train, given that no additional methods such as residual connections and normalization layers are adopted.

Finish the implementation of CNNtanh below, carefully following the specifications:

The model should have exactly 'k' many convolutional layers, followed by a linear (fully-connected) layer that actually outputs the logits for each of the 10 MNIST classes.

The network should consist of 3 stages, each with k/3 many convolutional layers (you can assume k is divisible by 3). Each conv layer should have a 3x3 kernel, a stride of 1 and a padding of 1 pixel (such that the output of the convolution has the same height and width as its input).

It should also have an average pooling layer at the end of each stage, with a 2x2 window (hence halving the spatial dimensions), and the number of channels should double from one stage to the other (starting with 4 in the first stage). Moreover, a Tanh activation should follow each convolution layer.

When k=3, for example, the network should be:

1. Stage 1 (1x28x28 input, 4x14x14 output):
    1. Conv layer with 1 input channel and 4 output channels, 3x3 kernel, stride=padding=1
    2. Tanh activation
    3. Average Pool with 2x2 kernel and stride 2
2. Stage 2 (4x14x14 input, 8x7x7 output):
    1. Conv layer with 4 input channels and 8 output channels, 3x3 kernel, stride=padding=1
    2. Tanh activation
    3. Average Pool with 2x2 kernel and stride 2
3. Stage 3 (8x7x7 input, 16x3x3 output):
    1. Conv layer with 8 input channels and 16 output channels, 3x3 kernel, stride=padding=1
    2. Tanh activation
    3. Average Pool with 2x2 kernel and stride 2
4. Fully-connected layer with 16 * 3 * 3=144 input dimension and 10 output dimension

Note that the model should not have any activation after the fully-connected layer: the PyTorch loss module that will be adopted takes logits as input and not class probabilities.

In contrast to the network exemplified above with k=3, when k=6 it should have two conv layers per stage instead of one (each one with a tanh activation following it).

Lastly, do not change the code block with a for loop in the end of init: its purpose to randomly initialize the parameters of the conv layers by sampling from a Gaussian with zero mean and 0.05 deviation.

In [11]:
class CNNtanh(nn.Module):
    def __init__(self, k):
        super(CNNtanh, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1

        k, lst, npt, otpt = (k//3)*3, [], 1, 4

        for i in range(0,3): #create 3 stages

          for _ in range(0, int(k/3)): #create k/3 convultion+tanh layers per stage, add to lst
            lst.extend([nn.Conv2d(npt, otpt, 3, 1, 1), nn.Tanh()])
            npt = otpt
     

          #at the end of stage i, double, channels, add pooling later
          lst.append(nn.AvgPool2d(2, stride=2))
          otpt = otpt *2

        self.mdl = nn.Sequential(*lst)

        sz = otpt//2 * 3* 3
        self.fc=nn.Linear(sz, 10)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                m.weight.data.normal_(0, 0.05)
                m.bias.data.zero_()
        
    def forward(self, input):

        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'
        u = self.mdl(input)
        u = u.view(-1, np.prod(u.size()[1:4]))
        u = self.fc(u)
        return u

The line below just instantiates the PyTorch Cross Entropy loss, whose inputs should be logits: hence the reason that the CNN should not have an activation after last (feedforward) layer.

In [12]:
criterion = torch.nn.CrossEntropyLoss()

Now, you should train CNNtanh with different values for k: your goal is to find the largest value for k such that the network achieves less than 20% error (either train or test) in 3 epochs. You should also choose an appropriate learning rate (but do not change the optimizer or the momentum settings!).

Note that CNNs can easily achieve under 2% test error on MNIST, but we're choosing 20% as a threshold since you will be training each network for only 3 epochs.

Remember to use values for k that are divisible by 3. When submitted, your notebook should have the training log of a network with two consecutive values for k (for example, 6 and 9) such that the network is 'trainable' with the smaller one but not 'trainable' with the larger one. It is fine for the training log to include runs with more than two values of k.

In [34]:
k=0
for i in range(1,100):
  k = k+3*(i//6+1)
  lr = .1

  print("Training Tanh CNN with {} layers".format(k))
  model = CNNtanh(k).cuda()
  optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
  train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader) 
  if test_errs >.20:  break

Training Tanh CNN with 3 layers
Epoch 000/003, Train Error 24.12% || Test Error 7.33%
Epoch 001/003, Train Error 5.86% || Test Error 4.10%
Epoch 002/003, Train Error 3.84% || Test Error 2.98%
Training Tanh CNN with 6 layers
Epoch 000/003, Train Error 54.64% || Test Error 7.55%
Epoch 001/003, Train Error 5.06% || Test Error 3.45%
Epoch 002/003, Train Error 3.16% || Test Error 2.75%
Training Tanh CNN with 9 layers
Epoch 000/003, Train Error 89.07% || Test Error 88.65%
Epoch 001/003, Train Error 88.73% || Test Error 65.23%
Epoch 002/003, Train Error 13.30% || Test Error 4.38%
Training Tanh CNN with 12 layers
Epoch 000/003, Train Error 89.04% || Test Error 88.65%
Epoch 001/003, Train Error 88.96% || Test Error 88.65%
Epoch 002/003, Train Error 88.95% || Test Error 88.65%


### Better Initialization

Next, we will change the initialization of the conv layers and see how it affects the trainability of deep networks. Instead of sampling from a Gaussian with a deviation of 0.05, you should sample from a Gaussian with a deviation $\sigma = \sqrt{\frac{1}{k^2 \cdot C_{out}}}$ or $\sigma = \sqrt{\frac{1}{k^2 \cdot C_{in}}}$, where $k$ is the kernel size ($k=3$ for 3x3 convolutions), $C_{in}$ is the number of input channels, and $C_{out}$ the number of output channels.

The model below should be exactly like CNNtanh except for the standard deviation of the normal distribution used to initialize the conv layers.

The paper 'Understanding the difficulty of training deep feedforward neural networks' by Glorot and Bengio provides some intuition behind such a choice for $\sigma$.

In [9]:
class CNNtanh_newinit(nn.Module):
    def __init__(self, k):
        super(CNNtanh_newinit, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1
        k, lst, npt, otpt = (k//3)*3, [], 1, 4

        for i in range(0,3): #create 3 stages
          for _ in range(0, int(k/3)): #create k/3 convultion+tanh layers per stage, add to lst
            lst.extend([nn.Conv2d(npt, otpt, 3, 1, 1), nn.Tanh()])
            npt = otpt
     

          #at the end of stage i, double, channels, add pooling layer
          lst.append(nn.AvgPool2d(2, stride=2))
          otpt = otpt *2

        self.mdl = nn.Sequential(*lst)

        sz = otpt//2 * 3* 3
        self.fc=nn.Linear(sz, 10)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                # insert code to compute sigma
                sigma = math.sqrt(1/(9*m.out_channels))
                m.weight.data.normal_(0, sigma)
                m.bias.data.zero_()
        
    def forward(self, input):

        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'
        u = self.mdl(input)
        u = u.view(-1, np.prod(u.size()[1:4]))
        u = self.fc(u)
        return u

Repeat the procedure of finding the maximum number of layers such that the network is still trainable, this time with CNNtanhinit.

In [35]:
k=0
for i in range(1,100):
  k = k+3*(i//6+1)
  lr = .01

  print("\nTraining Tanh CNN + new init with {} layers".format(k))
  model = CNNtanh_newinit(k).cuda()
  optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
  train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
  if test_errs >.20:  break


Training Tanh CNN + new init with 3 layers
Epoch 000/003, Train Error 42.33% || Test Error 20.26%
Epoch 001/003, Train Error 17.71% || Test Error 14.09%
Epoch 002/003, Train Error 13.24% || Test Error 10.78%

Training Tanh CNN + new init with 6 layers
Epoch 000/003, Train Error 35.02% || Test Error 15.47%
Epoch 001/003, Train Error 12.64% || Test Error 9.59%
Epoch 002/003, Train Error 8.68% || Test Error 6.82%

Training Tanh CNN + new init with 9 layers
Epoch 000/003, Train Error 46.74% || Test Error 14.95%
Epoch 001/003, Train Error 13.60% || Test Error 8.34%
Epoch 002/003, Train Error 7.32% || Test Error 5.45%

Training Tanh CNN + new init with 12 layers
Epoch 000/003, Train Error 38.38% || Test Error 14.09%
Epoch 001/003, Train Error 10.74% || Test Error 6.96%
Epoch 002/003, Train Error 6.28% || Test Error 4.76%

Training Tanh CNN + new init with 15 layers
Epoch 000/003, Train Error 40.82% || Test Error 12.36%
Epoch 001/003, Train Error 8.29% || Test Error 5.18%
Epoch 002/003, Trai

In [38]:
#Try to find exact k by going back through with smaller steps
k =  k-3*(i//6+1)
print(f'\nStarting back at k={k}')
for i in range(0,100):
  k = k+3*(i//6+1)
  lr = .01

  print("\nTraining Tanh CNN + new init with {} layers".format(k))
  model = CNNtanh_newinit(k).cuda()
  optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
  train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
  if test_errs >.20:  break


Starting back at k=87

Training Tanh CNN + new init with 90 layers
Epoch 000/003, Train Error 88.79% || Test Error 88.65%
Epoch 001/003, Train Error 82.06% || Test Error 52.27%
Epoch 002/003, Train Error 30.66% || Test Error 14.08%

Training Tanh CNN + new init with 93 layers
Epoch 000/003, Train Error 61.07% || Test Error 14.18%
Epoch 001/003, Train Error 9.53% || Test Error 6.42%
Epoch 002/003, Train Error 4.70% || Test Error 5.19%

Training Tanh CNN + new init with 96 layers
Epoch 000/003, Train Error 83.73% || Test Error 67.98%
Epoch 001/003, Train Error 41.26% || Test Error 20.96%
Epoch 002/003, Train Error 17.91% || Test Error 11.57%

Training Tanh CNN + new init with 99 layers
Epoch 000/003, Train Error 85.18% || Test Error 57.39%
Epoch 001/003, Train Error 42.42% || Test Error 23.09%
Epoch 002/003, Train Error 18.06% || Test Error 11.63%

Training Tanh CNN + new init with 102 layers
Epoch 000/003, Train Error 88.86% || Test Error 88.65%
Epoch 001/003, Train Error 86.14% || Tes

### CNN with ELU activations

In this section you should replace the Tanh activations of the previous network for Exponential Linear Units (ELUs). Complete CNNelu below, which should be exactly like CNNtanhinit except for ELU activations instead of Tanh (ELUs are readily available in PyTorch, check its documentation for more details).

In [36]:
class CNNelu(nn.Module):
    def __init__(self, k):
        super(CNNelu, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1

        k, lst, npt, otpt = (k//3)*3, [], 1, 4

        for i in range(0,3): #create 3 stages
          for _ in range(0, int(k/3)): #create k/3 convultion+tanh layers per stage, add to lst
            lst.extend([nn.Conv2d(npt, otpt, 3, 1, 1), nn.ELU()])
            npt = otpt
     

          #at the end of stage i, double, channels, add pooling layer
          lst.append(nn.AvgPool2d(2, stride=2))
          otpt = otpt *2

        self.mdl = nn.Sequential(*lst)

        sz = otpt//2 * 3* 3
        self.fc=nn.Linear(sz, 10)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                # insert code to compute sigma
                sigma = math.sqrt(1/(9*m.out_channels))
                m.weight.data.normal_(0, sigma)
                m.bias.data.zero_()
        
    def forward(self, input):

        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'
        u = self.mdl(input)
        u = u.view(-1, np.prod(u.size()[1:4]))
        u = self.fc(u)
        return u

Repeat the procedure of finding the maximum number of layers such that the network is still trainable, this time with CNNelu.

In [41]:
  k=0
  for i in range(1,100):
    k = k+3*(i//6+1)    
    lr = .01 

    print("\nTraining ELU CNN, with {} layers".format(k))
    model = CNNelu(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.20:  break


#Try to find exact k by going back through with smaller steps
  k =  k-3*(i//6+1)
  print(f'\nStarting back at k={k} and proceeding with smaller steps.')
  for i in range(0,100):
    k = k+3*(i//6+1)

    print("\nTraining ELU CNN, with {} layers".format(k))
    model = CNNelu(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.20:  break



Training ELU CNN, with 3 layers
Epoch 000/003, Train Error 26.46% || Test Error 8.57%
Epoch 001/003, Train Error 7.51% || Test Error 6.04%
Epoch 002/003, Train Error 5.48% || Test Error 3.93%

Training ELU CNN, with 6 layers
Epoch 000/003, Train Error 30.68% || Test Error 8.38%
Epoch 001/003, Train Error 6.45% || Test Error 5.30%
Epoch 002/003, Train Error 4.46% || Test Error 3.85%

Training ELU CNN, with 9 layers
Epoch 000/003, Train Error 44.22% || Test Error 8.33%
Epoch 001/003, Train Error 5.69% || Test Error 3.75%
Epoch 002/003, Train Error 3.23% || Test Error 2.55%

Training ELU CNN, with 12 layers
Epoch 000/003, Train Error 33.47% || Test Error 5.83%
Epoch 001/003, Train Error 3.95% || Test Error 2.85%
Epoch 002/003, Train Error 2.56% || Test Error 1.99%

Training ELU CNN, with 15 layers
Epoch 000/003, Train Error 36.40% || Test Error 6.22%
Epoch 001/003, Train Error 4.36% || Test Error 3.60%
Epoch 002/003, Train Error 2.82% || Test Error 2.53%

Training ELU CNN, with 21 layers

### CNN with Batch Normalization

Next, you will check how batch normalization can make deep networks easier to train. Implement the network below, which should be exactly like CNNelu except for additional BatchNorm2d layers after each convolution (before the ELU activation).

Note that BatchNorm2d modules require the number of channels as argument -- see the PyTorch documentation for more details.

In [42]:
class CNNeluBN(nn.Module):
    def __init__(self, k):
        super(CNNeluBN, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1
        k, lst, npt, otpt = (k//3)*3, [], 1, 4

        for i in range(0,3): #create 3 stages
          for _ in range(0, int(k/3)): #create k/3 convultion+tanh layers per stage, add to lst
            lst.extend([nn.Conv2d(npt, otpt, 3, 1, 1), nn.BatchNorm2d(otpt), nn.ELU()]) 
            npt = otpt
     

          #at the end of stage i, double, channels, add pooling layer
          lst.append(nn.AvgPool2d(2, stride=2))
          otpt = otpt *2

        self.mdl = nn.Sequential(*lst)

        sz = otpt//2 * 3* 3
        self.fc=nn.Linear(sz, 10)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                # insert code to compute sigma
                sigma = math.sqrt(1/(9*m.out_channels))
                m.weight.data.normal_(0, sigma)
                m.bias.data.zero_()
        
    def forward(self, input):
        
        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'
        u = self.mdl(input)
        u = u.view(-1, np.prod(u.size()[1:4]))
        u = self.fc(u)
        return u

Repeat the procedure of finding the maximum number of layers such that the network is still trainable, this time with CNNeluBN.

In [44]:
  k=0
  for i in range(1,100):
    k = k+3*(i//6+1)    
    lr = .01  

    print("\nTraining ELU CNN + BN with {} layers".format(k))
    model = CNNeluBN(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = ResNet(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break 
      
#Try to find exact k by going back through with smaller steps
  k =  k-3*(i//6+1)
  print(f'\nStarting back at k={k} and proceeding with smaller steps.')
  for i in range(0,100):
    k = k+3*(i//6+1)

    print("\nTraining ELU CNN + BN with {} layers".format(k))
    model = CNNeluBN(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: break




Training ELU CNN + BN with 3 layers
Epoch 000/003, Train Error 13.00% || Test Error 3.88%
Epoch 001/003, Train Error 3.65% || Test Error 2.18%
Epoch 002/003, Train Error 2.30% || Test Error 1.87%

Training ELU CNN + BN with 6 layers
Epoch 000/003, Train Error 11.81% || Test Error 3.15%
Epoch 001/003, Train Error 3.17% || Test Error 1.77%
Epoch 002/003, Train Error 2.06% || Test Error 1.67%

Training ELU CNN + BN with 9 layers
Epoch 000/003, Train Error 9.33% || Test Error 2.10%
Epoch 001/003, Train Error 2.42% || Test Error 1.59%
Epoch 002/003, Train Error 1.73% || Test Error 1.29%

Training ELU CNN + BN with 12 layers
Epoch 000/003, Train Error 7.87% || Test Error 1.97%
Epoch 001/003, Train Error 12.50% || Test Error 4.12%
Epoch 002/003, Train Error 3.38% || Test Error 2.98%

Training ELU CNN + BN with 15 layers
Epoch 000/003, Train Error 8.44% || Test Error 2.58%
Epoch 001/003, Train Error 2.96% || Test Error 2.61%
Epoch 002/003, Train Error 1.88% || Test Error 1.55%

Training ELU C

I am restarting at 57, because it didnt follow the observed trend for the model to fail at that point.

In [56]:
  k=57
  print(f'\nStarting back at k={k} and proceeding with smaller steps.')
  for i in range(0,100):
    k = k+3*(i//6+1)

    print("\nTraining ELU CNN + BN with {} layers".format(k))
    model = CNNeluBN(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = ResNet(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break 


Starting back at k=57 and proceeding with smaller steps.

Training ELU CNN + BN with 60 layers
Epoch 000/003, Train Error 15.24% || Test Error 3.76%
Epoch 001/003, Train Error 3.73% || Test Error 2.39%
Epoch 002/003, Train Error 2.32% || Test Error 2.09%

Training ELU CNN + BN with 63 layers
Epoch 000/003, Train Error 15.37% || Test Error 4.28%
Epoch 001/003, Train Error 4.19% || Test Error 2.32%
Epoch 002/003, Train Error 2.66% || Test Error 2.23%

Training ELU CNN + BN with 66 layers
Epoch 000/003, Train Error 16.57% || Test Error 4.82%
Epoch 001/003, Train Error 15.17% || Test Error 4.78%
Epoch 002/003, Train Error 4.25% || Test Error 3.67%

Training ELU CNN + BN with 69 layers
Epoch 000/003, Train Error 19.56% || Test Error 4.51%
Epoch 001/003, Train Error 4.54% || Test Error 3.56%
Epoch 002/003, Train Error 2.89% || Test Error 2.25%

Training ELU CNN + BN with 72 layers
Epoch 000/003, Train Error 16.10% || Test Error 3.28%
Epoch 001/003, Train Error 23.95% || Test Error 6.45%
Epo

### Residual Networks

Finally, you experiment adding residual connections to a CNN.

To implement the model below, you should add a 'skip connection' to 'Conv->BatchNorm->ELU' blocks whenever the shape of the block's input and output are the same: this will be the case for every such block except for the first ones in each stage, as they double the number of channels.

More specifically, you should change $u = ELU(BatchNorm(Conv(x)))$ to $u = ELU(BatchNorm(Conv(x))) + x$, where $x$ and $u$ denote the block's input and output, respectively.

You should take your CNNeluBN implementation and add skip-connections as described above.

Note that there are key differences between the resulting model and the actual ResNet proposed by He et al. in 'Deep Residual Learning for Image Recognition', for example the use of ELU activations instead of ReLU and the exact position of skip-connections.

In [49]:
class ResNet(nn.Module):
    def __init__(self, k):
        super(ResNet, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1
        k, npt, otpt = (k//3)*3, 1, 4
        lst = nn.ModuleList()  


        for i in range(0,3): #create 3 stages
          for _ in range(0, int(k/3)): #create k/3 convultion blocks per stage, add to lst
            lst.extend([nn.Conv2d(npt, otpt, 3, 1, 1), nn.BatchNorm2d(otpt), nn.ELU()]) #be cautious with untested code
            npt = otpt
     

          #at the end of stage i, double, channels, add pooling layer
          lst.append(nn.AvgPool2d(2, stride=2))
          otpt = otpt *2

        self.mdl = lst

        sz = otpt//2 * 3* 3
        self.fc=nn.Linear(sz, 10)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                # insert code to compute sigma
                sigma = math.sqrt(1/(9*m.out_channels))
                m.weight.data.normal_(0, sigma)
                m.bias.data.zero_()
        
        
    def forward(self, input):
        
        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'
        u = input

        #do main part of model
        for layer in self.mdl:
          if isinstance(layer, nn.Conv2d): 
            u_block = u #save input when first encounter a new block
            skip = layer.in_channels == layer.out_channels
          if isinstance(layer, nn.ELU) and skip: u = layer(u)+u_block #when in == out channels and is ELU, do skip
          else: u = layer(u)

        #do fully connected  
        u = u.view(-1, np.prod(u.size()[1:4]))
        u = self.fc(u)
        return u

Repeat the procedure of finding the maximum number of layers such that the network is still trainable, this time with the 'ResNet' model.

In [52]:
  k=0
  for i in range(1,100):
    k = k+3*(i//6+1)    
    lr = .01  

    print("\nTraining ResNet with {} layers".format(k))
    model = ResNet(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = ResNet(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break 
      
#Try to find exact k by going back through with smaller steps
  k =  k-3*(i//6+1)
  print(f'\nStarting back at k={k} and proceeding with smaller steps.')
  for i in range(0,100):
    k = k+3*(i//6+1)

    print("\nTraining ResNet with {} layers".format(k))
    model = ResNet(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: break





Training ResNet with 3 layers
Epoch 000/003, Train Error 15.85% || Test Error 4.86%
Epoch 001/003, Train Error 4.23% || Test Error 2.59%
Epoch 002/003, Train Error 2.76% || Test Error 2.01%

Training ResNet with 6 layers
Epoch 000/003, Train Error 10.08% || Test Error 2.78%
Epoch 001/003, Train Error 88.12% || Test Error 90.20%
Epoch 002/003, Train Error 90.13% || Test Error 90.20%
Epoch 000/003, Train Error 10.80% || Test Error 2.86%
Epoch 001/003, Train Error 3.57% || Test Error 2.17%
Epoch 002/003, Train Error 1.96% || Test Error 1.92%

Training ResNet with 9 layers
Epoch 000/003, Train Error 8.79% || Test Error 3.03%
Epoch 001/003, Train Error 2.62% || Test Error 1.94%
Epoch 002/003, Train Error 1.80% || Test Error 1.51%

Training ResNet with 12 layers
Epoch 000/003, Train Error 8.51% || Test Error 2.22%
Epoch 001/003, Train Error 2.62% || Test Error 1.99%
Epoch 002/003, Train Error 1.79% || Test Error 1.50%

Training ResNet with 15 layers
Epoch 000/003, Train Error 6.92% || Test 

In [55]:
  k =  51
  print(f'\nStarting back at k={k} and proceeding with smaller steps.')
  for i in range(0,100):
    k = k+3*(i//6+1)

    print("\nTraining ResNet with {} layers".format(k))
    model = ResNet(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = ResNet(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break 



Starting back at k=51 and proceeding with smaller steps.

Training ResNet with 54 layers
Epoch 000/003, Train Error 7.72% || Test Error 2.33%
Epoch 001/003, Train Error 2.62% || Test Error 1.58%
Epoch 002/003, Train Error 1.65% || Test Error 1.68%

Training ResNet with 57 layers
Epoch 000/003, Train Error 8.78% || Test Error 2.15%
Epoch 001/003, Train Error 2.51% || Test Error 1.58%
Epoch 002/003, Train Error 1.63% || Test Error 1.40%

Training ResNet with 60 layers
Epoch 000/003, Train Error 9.39% || Test Error 2.35%
Epoch 001/003, Train Error 89.05% || Test Error 90.20%
Epoch 002/003, Train Error 90.13% || Test Error 90.20%
Epoch 000/003, Train Error 7.84% || Test Error 2.43%
Epoch 001/003, Train Error 2.54% || Test Error 1.77%
Epoch 002/003, Train Error 1.63% || Test Error 1.35%

Training ResNet with 63 layers
Epoch 000/003, Train Error 10.43% || Test Error 2.45%
Epoch 001/003, Train Error 2.50% || Test Error 1.74%
Epoch 002/003, Train Error 1.67% || Test Error 1.51%

Training ResN

**<font color='blue'>
    Summarize your results and observations regarding the experiments above. What was the maximum number of layers for each of the five models such that training remained successful? Briefly discuss why you think each modification helped/harmed the trainability of deep models.
</font>**

**<font color='red'> --------------------------------------------------------------------- ANSWER (BEGIN) ---------------------------------------------------------------------
</font>**

## Short Answer Responses

### CNN w/Tanh Activations
  - The max layer that I reached was 12. 

### CNN w/Tanh Activations + Better Initialization
  -  When I added the suggested initalization, the max k went up to 96 or 102. This is a significant jump. Since each layer was initalized in a way that took in account its own dimensions as opposed to just being random, I would expect to see much better results.
  
### CNN w/ Better Activation + ELU Activation
  -  This architecture made it to around k=90. So, it actually did worse than the tanh activation. It is not clear to me, why this happened as ReLu seems better regarded in the ML community. Perhaps, something about this architecture generated negative values and forced our gradients to 0. 
  
I believe this also makes the architecture very sensitive to randomness of the initial values. In my experiments for several archetectures, it can be seen that in a range of k that was doing fine before, it suddenly jumps to 80+% error, but if reran it went down to something more expected. 

### CNN w/... + Batch Normalizations
- This architecture actually didn't fail at the 20% error threshold consistently until around 220 layers but it was clearly struggling starting at 160. This is much higher than other architectures. It is particularly interesting that even with ELU activation which did noticeably worse previously, the batch normalization allowed a deeper depth than I had seen up until this point. Perhaps the normalization process allowed the gradients to stay within an reasonable range. 

### CNN w/... + Skip Connections
  - This architecure made it higher than most of the others, maxing out somewhere around 160 layers. It was subject to the randomness I mentioned earlier. Given that the batch normalization also performed similarly, it is not clear if the skip connections helped any. 


**<font color='red'> ---------------------------------------------------------------------- ANSWER (END) ----------------------------------------------------------------------
</font>**

### Interactions: Batch Norm and Initialization

Intuitively, batch norm should make the model more robust to changes in the magnitude of the network's weights: informally, scaling up all the elements of a conv layer's filters by a factor of 10 would not affect the network's output as long as there is a batch norm layer following such convolution, as the normalization would undo the scaling.

To check how this intuition translates to practical settings, you should change the original 'CNNtanh' model so that it incorporates batch norm layers (like you have done when modifying 'CNNelu' into 'CNNeluBN').

The model below should adopt the naive initialization procedure of sampling from a Gaussian with a deviation of 0.05, not the more sophisticated one that you implemented previously

In [59]:
class CNNtanhBN_oldinit(nn.Module):
    def __init__(self, k):
        super(CNNtanhBN_oldinit, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1
        k, lst, npt, otpt = (k//3)*3, [], 1, 4

        for i in range(0,3): #create 3 stages

          for _ in range(0, int(k/3)): #create k/3 convultion+tanh layers per stage, add to lst
            lst.extend([nn.Conv2d(npt, otpt, 3, 1, 1), nn.BatchNorm2d(otpt), nn.Tanh()])
            npt = otpt
     

          #at the end of stage i, double, channels, add pooling later
          lst.append(nn.AvgPool2d(2, stride=2))
          otpt = otpt *2

        self.mdl = nn.Sequential(*lst)

        sz = otpt//2 * 3* 3
        self.fc=nn.Linear(sz, 10)


        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                m.weight.data.normal_(0, 0.05)
                m.bias.data.zero_()
        
    def forward(self, input):

        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'
        u = self.mdl(input)
        u = u.view(-1, np.prod(u.size()[1:4]))
        u = self.fc(u)        
        return u

Repeat the procedure of finding the maximum number of layers such that the network is still trainable, this time with CNNeluBN_oldinit.

In [61]:
  k=0
  for i in range(1,100):
    k = k+3*(i//6+1)    
    lr = .01  

    print("\nTraining Tanh CNN + BN + naive init with {} layers".format(k))
    model = CNNtanhBN_oldinit(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = CNNtanhBN_oldinit(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break 
      
#Try to find exact k by going back through with smaller steps
  k =  k-3*(i//6+1)
  print(f'\nStarting back at k={k} and proceeding with smaller steps.')
  for i in range(0,100):
    k = k+3*(i//6+1)

    print("\nTraining Tanh CNN + BN + naive init with {} layers".format(k))
    model = CNNtanhBN_oldinit(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = CNNtanhBN_oldinit(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break s


Training Tanh CNN + BN + naive init with 3 layers
Epoch 000/003, Train Error 15.77% || Test Error 5.23%
Epoch 001/003, Train Error 9.33% || Test Error 5.27%
Epoch 002/003, Train Error 4.71% || Test Error 3.85%

Training Tanh CNN + BN + naive init with 6 layers
Epoch 000/003, Train Error 12.37% || Test Error 6.56%
Epoch 001/003, Train Error 10.71% || Test Error 3.96%
Epoch 002/003, Train Error 3.85% || Test Error 2.98%

Training Tanh CNN + BN + naive init with 9 layers
Epoch 000/003, Train Error 10.10% || Test Error 3.58%
Epoch 001/003, Train Error 8.77% || Test Error 3.12%
Epoch 002/003, Train Error 2.79% || Test Error 2.17%

Training Tanh CNN + BN + naive init with 12 layers
Epoch 000/003, Train Error 10.33% || Test Error 3.04%
Epoch 001/003, Train Error 9.14% || Test Error 3.73%
Epoch 002/003, Train Error 3.35% || Test Error 2.75%

Training Tanh CNN + BN + naive init with 15 layers
Epoch 000/003, Train Error 10.29% || Test Error 2.95%
Epoch 001/003, Train Error 15.68% || Test Error 

**<font color='blue'>
    Compare CNNtanh (model with naive initialization and no batch norm), CNNtanh_newinit (model with better initialization and no batch norm), and CNNtanhBN_oldinit (model with naive initialization and batch norm), in terms of how deep each could be while being trainable, and discuss your thoughts one how batch norm interacts with the way parameters are initialized.
</font>**

**<font color='red'> --------------------------------------------------------------------- ANSWER (BEGIN) ---------------------------------------------------------------------
</font>**

CNNtanh only made it to 12 layers. CNNtanh_newinit to 102, and CNNtanhBN_oldinit to around 100. It seem that even with a random initialization the batch normalization improves the architecture greatly. 




**<font color='red'> ---------------------------------------------------------------------- ANSWER (END) ----------------------------------------------------------------------
</font>**

### Interactions: Batch Norm and Residual Connections

Lastly, implement and train a CNN with residual connections but without batch normalization layers -- the goal here is to check how residuals interact with normalization.

The model below should be exactly like ResNet, except that it should not have batch norm layers.

In [62]:
class ResNet_noBN(nn.Module):
    def __init__(self, k):
        super(ResNet_noBN, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1

        k, npt, otpt = (k//3)*3, 1, 4
        lst = nn.ModuleList()  


        for i in range(0,3): #create 3 stages
          for _ in range(0, int(k/3)): #create k/3 convultion blocks per stage, add to lst
            lst.extend([nn.Conv2d(npt, otpt, 3, 1, 1), nn.ELU()]) #be cautious with untested code
            npt = otpt
     

          #at the end of stage i, double, channels, add pooling layer
          lst.append(nn.AvgPool2d(2, stride=2))
          otpt = otpt *2

        self.mdl = lst

        sz = otpt//2 * 3* 3
        self.fc=nn.Linear(sz, 10)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                # insert code to compute sigma
                sigma = math.sqrt(1/(9*m.out_channels))
                m.weight.data.normal_(0, sigma)
                m.bias.data.zero_()
        
        
    def forward(self, input):
        
        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'
        u = input

        #do main part of model
        for layer in self.mdl:
          if isinstance(layer, nn.Conv2d): 
            u_block = u #save input when first encounter a new block
            skip = layer.in_channels == layer.out_channels
          if isinstance(layer, nn.ELU) and skip: u = layer(u)+u_block #when in == out channels and is ELU, do skip
          else: u = layer(u)

        #do fully connected  
        u = u.view(-1, np.prod(u.size()[1:4]))
        u = self.fc(u)
        return u

Repeat the procedure of finding the maximum number of layers such that the network is still trainable, this time with ResNet_noBN.

In [63]:
  k=0
  for i in range(1,100):
    k = k+3*(i//6+1)    
    lr = .01  

    print("\nTraining ResNet w/o BN with {} layers".format(k))
    model = ResNet_noBN(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = ResNet_noBN(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break 
      
#Try to find exact k by going back through with smaller steps
  k =  k-3*(i//6+1)
  print(f'\nStarting back at k={k} and proceeding with smaller steps.')
  for i in range(0,100):
    k = k+3*(i//6+1)

    print("\nTraining ResNet w/o BN with {} layers".format(k))
    model = ResNet_noBN(k).cuda()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
    train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
    if test_errs >.2: 
      #try again
      model = ResNet_noBN(k).cuda()
      optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
      train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)
      if test_errs >.2: break 




Training ResNet w/o BN with 3 layers
Epoch 000/003, Train Error 43.90% || Test Error 12.81%
Epoch 001/003, Train Error 8.96% || Test Error 6.31%
Epoch 002/003, Train Error 5.80% || Test Error 4.98%

Training ResNet w/o BN with 6 layers
Epoch 000/003, Train Error 21.69% || Test Error 5.43%
Epoch 001/003, Train Error 4.13% || Test Error 3.09%
Epoch 002/003, Train Error 2.89% || Test Error 2.14%

Training ResNet w/o BN with 9 layers
Epoch 000/003, Train Error 12.21% || Test Error 3.26%
Epoch 001/003, Train Error 3.11% || Test Error 2.05%
Epoch 002/003, Train Error 2.20% || Test Error 1.91%

Training ResNet w/o BN with 12 layers
Epoch 000/003, Train Error 11.60% || Test Error 4.32%
Epoch 001/003, Train Error 2.51% || Test Error 1.94%
Epoch 002/003, Train Error 1.83% || Test Error 1.43%

Training ResNet w/o BN with 15 layers
Epoch 000/003, Train Error 11.10% || Test Error 2.35%
Epoch 001/003, Train Error 2.23% || Test Error 1.97%
Epoch 002/003, Train Error 1.63% || Test Error 1.56%

Traini

**<font color='blue'>
    Compare ResNet and ResNet_noBN in terms of how deep each could be while being trainable, and discuss your thoughts one how batch norm interacts with residual connections.
</font>**

**<font color='red'> --------------------------------------------------------------------- ANSWER (BEGIN) ---------------------------------------------------------------------
</font>**

ResNet_noBN did very poorly maxing at around 15 layers compared to Resnet at around 160. I think that batch norming is making all the difference and something about the architecure is not allowing the residual connections to improve very much on our base architecture. 

**<font color='red'> ---------------------------------------------------------------------- ANSWER (END) ----------------------------------------------------------------------
</font>**

### (Optional) Multiple Loss Heads

In this optional section, your goal is to incorporate the idea of having multiple loss heads throughout the network, distributed across its depth.

For the CNNelu_multihead model below, you should take the CNNelu model that you implemented previously and add two additional classification heads, connected to the outputs of stages 1 and 2.

More specifically, the outputs of stages 1 and 2, with shapes 4x14x14 and 8x7x7, should be connected to new fully-connected layers that map them to a 10-dimensional vector (logits for the 10 MNIST classes). The network should output three logit vectors (the original one at the end of the network plus the two new ones) instead of just one, and the loss should be computed as the average of the cross entropies between the true target and each of the three predictions.

Note that you will likely have to change the implementation of train_epoch() and test() to accomodate the fact that this model will output three logit vectors instead of one.

In [ ]:
class CNNelu_multihead(nn.Module):
    def __init__(self, k):
        super(CNNelu_multihead, self).__init__()

        # write code here to instantiate layers
        # for example, self.conv = nn.Conv2d(1, 4, 3, 1, 1)
        # creates a conv layer with 1 input channel, 4 output
        # channels, a 3x3 kernel, and stride=padding=1

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                # insert code to compute sigma
                sigma = 
                m.weight.data.normal_(0, sigma)
                m.bias.data.zero_()
        
    def forward(self, input):
        # write code here to define how the output u is computed
        # from the input and the model's layers
        # for example, u = self.conv(input) defines u
        # to be simply the output of self.conv given 'input'

        return u1, u2, u3

Repeat the procedure of finding the maximum number of layers such that the network is still trainable, this time with CNNelu_multihead.

In [ ]:
k = 
lr = 

print("\nTraining ELU CNN + multiloss with {} layers".format(k))
model = CNNelu_multihead(k).cuda()
optimizer = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
train_errs, test_errs = train(3, model, criterion, optimizer, train_loader, test_loader)

**<font color='blue'>
    Did the adoption of multiple loss heads help train deeper models? How did it compare to the adoption of batch normalization, in terms of how deeper each of the two approaches enabled the network to be while staying trainable?
</font>**

**<font color='red'> --------------------------------------------------------------------- ANSWER (BEGIN) ---------------------------------------------------------------------
</font>**

**<font color='red'> ---------------------------------------------------------------------- ANSWER (END) ----------------------------------------------------------------------
</font>**